storage/src/main/java/io/delta/storage/GCSLogStore.java

```diff
@@ -0,0 +1,189 @@
+                            String.format(
+                                    "... run in separate thread using $s",
+                                    GCSLogStore.class.getSimpleName(),
+                                    " static method runInNewThread"),
+                            "",
+                            "",
+                            -1)
+            );
+            StackTraceElement[] finalStackTrace = Stream.concat(
+                    Stream.concat(baseStackTrace,
+                            placeHolderStackElem),
+                    baseStackTrace).toArray(StackTraceElement[]::new);
+            realException.setStackTrace(finalStackTrace);
+            throw realException;
+        } else {
+            return resultHolder.get(0);
+        }
+    }
+
+    @Override
+    public void write(Path path,
+                      Iterator<String> actions,
+                      Boolean overwrite,
+                      Configuration hadoopConf) throws IOException {
+        FileSystem fs = path.getFileSystem(hadoopConf);
+
+        // This is needed for the tests to throw error with local file system.
+        if (fs instanceof LocalFileSystem && !overwrite && fs.exists(path)) {
+            throw new FileAlreadyExistsException(path.toString());
+        }
+
+        // GCS may upload an incomplete file when the current thread is interrupted, hence we move
+        // the write to a new thread so that the write cannot be interrupted.
+        // TODO Remove this hack when the GCS Hadoop connector fixes the issue.
+        // If overwrite=false and path already exists, gcs-connector will throw
+        // org.apache.hadoop.fs.FileAlreadyExistsException after fs.create is invoked.
+        // This should be mapped to java.nio.file.FileAlreadyExistsException.
+        FSDataOutputStream stream = fs.create(path, overwrite);
+
+        Callable body = () -> {
+            try {
+                while (actions.hasNext()) {
+                    stream.write((actions.next() + "\n").getBytes(StandardCharsets.UTF_8));
+                }
+            } finally {
+                stream.close();
+
+            }
+            return "";
+        };
+
+        try {
+            runInNewThread("delta-gcs-logstore-write", true, body);
+        } catch (org.apache.hadoop.fs.FileAlreadyExistsException e) {
+            throw new FileAlreadyExistsException(path.toString());
+            // GCS uses preconditions to handle race conditions for multiple writers.
+            // If path gets created between fs.create and stream.close by an external
+            // agent or race conditions. Then this block will execute.
+            // Reference: https://cloud.google.com/storage/docs/generations-preconditions
+        } catch (IOException e) {
+            if (isPreconditionFailure(e) && !overwrite)
+                throw new FileAlreadyExistsException(path.toString());
+        } catch (Throwable throwable) {
```

Comment:
do we need this `catch` statement?

`GCSLogStore.scala` doesn't seem to have this.

Reply:
Avoided this by removing the .initCause(e) (mentioned in the above comment)